components/cop_codegen/src/rpn_function.rs

```diff
@@ -62,3 +71,595 @@ impl RpnFunctionOpts {
+        let (impl_generics, _, where_clause) = generics.split_for_impl();
+        let (ctx_type, payload_type, result_type) = (ctx_type(), payload_type(), result_type());
+        quote! {
+            impl #impl_generics #fn_trait_ident #ty_generics for #tp_ident #where_clause {
+                default fn eval(
+                    self,
+                    rows: usize,
+                    ctx: &mut #ctx_type,
+                    payload: #payload_type,
+                ) -> #result_type {
+                    panic!("Cannot apply {} on {:?}", #fn_name, self)
+                }
+            }
+        }
+    }
+
+    fn generate_real_fn_trait_impl(&self) -> TokenStream {
+        let mut generics = self.item_fn.decl.generics.clone();
+        generics
+            .params
+            .push(parse_str::<GenericParam>("'arg_").unwrap());
+        let mut tp = NULL.to_string();
+        for (arg_index, arg_type) in self.arg_types.iter().enumerate().rev() {
+            let arg_name = format!("Arg{}_", arg_index);
+            let generic_param = format!(
+                "{}: {}<Type = &'arg_ Option<{}>>",
+                arg_name,
+                RPN_FN_ARG,
+                arg_type.into_token_stream()
+            );
+            generics.params.push(parse_str(&generic_param).unwrap());
+            tp = format!("{}<{}, {}>", ARG, arg_name, tp);
+        }
+        let fn_ident = &self.item_fn.ident;
+        let fn_trait_ident = &self.fn_trait_ident;
+        let tp = parse_str::<Type>(&tp).unwrap();
+        let (_, ty_generics, _) = self.item_fn.decl.generics.split_for_impl();
+        let (impl_generics, _, where_clause) = generics.split_for_impl();
+        let mut extract_args = String::new();
+        let mut call_args = String::new();
+        for meta in &self.meta {
+            call_args += &format!("{}, ", meta);
+        }
+        for arg_index in 0..self.arg_types.len() {
+            let arg_var = format!("arg{}", arg_index);
+            extract_args += &format!("let ({}, arg) = arg.extract(row);", arg_var);
+            call_args += &format!("{}, ", arg_var);
+        }
+        let extract_args: TokenStream = extract_args.parse().unwrap();
+        let call_args: TokenStream = call_args.parse().unwrap();
+        let ty_generics_turbofish = ty_generics.as_turbofish();
+        let (ctx_type, payload_type, result_type) = (ctx_type(), payload_type(), result_type());
+        quote! {
+            impl #impl_generics #fn_trait_ident #ty_generics for #tp #where_clause {
+                fn eval(
+                    self,
+                    rows: usize,
+                    ctx: &mut #ctx_type,
+                    payload: #payload_type,
+                ) -> #result_type {
+                    let arg = &self;
+                     let mut result = Vec::with_capacity(rows);
+                     for row in 0..rows {
+                         #extract_args
```

Comment:
We can remove L185-L196 and generate `extract_args` and `call_args` in `quote!` directly. See: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9f97453d8f5c10370b82d757bd8d9ace

Reply:
Good idea!